crates/matrix-sdk-base/src/deserialized_responses.rs

```diff
@@ -61,81 +63,200 @@ pub struct MembersResponse {
+pub enum SyncOrStrippedState<C>
+where
+    C: StaticStateEventContent + RedactContent,
+    C::Redacted: RedactedStateEventContent + fmt::Debug + Clone,
+{
+    /// An event from a room in joined or left state.
+    Sync(SyncStateEvent<C>),
+    /// An event from a room in invited state.
+    Stripped(StrippedStateEvent<C::PossiblyRedacted>),
+}
+
+impl<C> SyncOrStrippedState<C>
+where
+    C: StaticStateEventContent + RedactContent,
+    C::Redacted: RedactedStateEventContent<StateKey = C::StateKey>
+        + RedactedStateEventContent
+        + fmt::Debug
+        + Clone,
+    C::PossiblyRedacted: PossiblyRedactedStateEventContent<StateKey = C::StateKey>,
+{
+    /// If this is a `SyncStateEvent`, return a reference to the inner event.
+    pub fn as_sync(&self) -> Option<&SyncStateEvent<C>> {
+        match self {
+            Self::Sync(ev) => Some(ev),
+            Self::Stripped(_) => None,
+        }
+    }
+
+    /// If this is a `StrippedStateEvent`, return a reference to the inner
+    /// event.
+    pub fn as_stripped(&self) -> Option<&StrippedStateEvent<C::PossiblyRedacted>> {
+        match self {
+            Self::Sync(_) => None,
+            Self::Stripped(ev) => Some(ev),
         }
     }
 
     /// The sender of this event.
     pub fn sender(&self) -> &UserId {
         match self {
-            MemberEvent::Sync(e) => e.sender(),
-            MemberEvent::Stripped(e) => &e.sender,
+            Self::Sync(e) => e.sender(),
+            Self::Stripped(e) => &e.sender,
         }
     }
 
     /// The ID of this event.
     pub fn event_id(&self) -> Option<&EventId> {
         match self {
-            MemberEvent::Sync(e) => Some(e.event_id()),
-            MemberEvent::Stripped(_) => None,
+            Self::Sync(e) => Some(e.event_id()),
+            Self::Stripped(_) => None,
         }
     }
 
-    /// The Server Timestamp of this event.
+    /// The server timestamp of this event.
     pub fn origin_server_ts(&self) -> Option<MilliSecondsSinceUnixEpoch> {
         match self {
-            MemberEvent::Sync(e) => Some(e.origin_server_ts()),
-            MemberEvent::Stripped(_) => None,
+            Self::Sync(e) => Some(e.origin_server_ts()),
+            Self::Stripped(_) => None,
         }
     }
 
-    /// The membership state of the user
+    /// The state key associated to this state event.
+    pub fn state_key(&self) -> &C::StateKey {
+        match self {
+            Self::Sync(e) => e.state_key(),
+            Self::Stripped(e) => &e.state_key,
+        }
+    }
+}
+
+/// Wrapper around both MemberEvent-Types
+pub type MemberEvent = SyncOrStrippedState<RoomMemberEventContent>;
+
+impl MemberEvent {
+    /// The inner Content of the wrapped Event.
+    pub fn original_content(&self) -> Option<&RoomMemberEventContent> {
+        match self {
+            MemberEvent::Sync(e) => e.as_original().map(|e| &e.content),
+            MemberEvent::Stripped(e) => Some(&e.content),
+        }
+    }
```

Comment:
Couldn't this be a function on `SyncOrStrippedState<C>` for any `C` (with appropriate bounds)?